test_tools/log_parser/joiner.py

```diff
@@ -0,0 +1,194 @@
+        self.file.write(struct.pack('I', len(payload)))
+        self.file.write(payload)
+        if padding_bytes > 0:
+            self.file.write(bytes([0] * padding_bytes))
+
+    def write_header(self, properties):
+        self.file.write(b'VWFB')
+        self.file.write(struct.pack('I', 1))
+
+        builder = flatbuffers.Builder(0)
+        kv_offsets = []
+        for key in properties:
+            value = properties[key]
+            k_off = builder.CreateString(str(key))
+            v_off = builder.CreateString(str(value))
+            KeyValueStart(builder)
+            KeyValueAddKey(builder, k_off)
+            KeyValueAddValue(builder, v_off)
+            kv_offsets.append(KeyValueEnd(builder))
+
+        props_off = mk_offsets_vector(builder, kv_offsets, FileHeaderStartPropertiesVector)
+
+        FileHeaderStart(builder)
+        FileHeaderAddJoinTime(builder, mk_timestamp(builder))
+        FileHeaderAddProperties(builder, props_off)
+
+        header_off = FileHeaderEnd(builder)
+        builder.Finish(header_off)
+        self.write_message(MSG_TYPE_HEADER, builder.Output())
+
+    def write_join_msg(self, events):
+        builder = flatbuffers.Builder(0)
+
+        evt_offsets = []
+        for evt in events:
+            payload_off = mk_bytes_vector(builder, evt)
+            JoinedEventStart(builder)
+            JoinedEventAddEvent(builder, payload_off)
+            JoinedEventAddTimestamp(builder, mk_timestamp(builder))
+            evt_offsets.append(JoinedEventEnd(builder))
+
+
+        evt_array_offset = mk_offsets_vector(builder, evt_offsets, JoinedPayloadStartEventsVector)
+
+        JoinedPayloadStart(builder)
+        JoinedPayloadAddEvents(builder, evt_array_offset)
+        joined_payload_off = JoinedPayloadEnd(builder)
+
+        builder.Finish(joined_payload_off)
+        self.write_message(MSG_TYPE_REGULAR, builder.Output())
+
+    def write_eof(self):
+        self.write_message(MSG_TYPE_EOF, b'')
+
+interactions_file = PreambleStreamReader('cb_v2.fb')
+observations_file = PreambleStreamReader('f-reward_v2.fb')
+
+result_file = 'merged.log'
+
+def get_event_id(ser_evt):
+    evt = Event.GetRootAsEvent(ser_evt.PayloadAsNumpy(), 0)
+    m = evt.Meta()
+    if m:
+        return m.Id()
+    return None
+
+
+observations = dict()
+obs_count = 0
+obs_ids = 0
+for msg in observations_file.messages():
+    batch = EventBatch.GetRootAsEventBatch(msg, 0)
+    for i in range(0, batch.EventsLength()):
+        ser_evt = batch.Events(i)
+        evt_id = get_event_id(ser_evt)
+        if evt_id not in observations:
+            observations[evt_id] = []
+            obs_ids += 1
+        observations[evt_id].append(ser_evt.PayloadAsNumpy())
+        obs_count +=1
+
+print(f'found {obs_count} observations with {obs_ids} ids')
+
+bin_f = BinLogWriter(result_file)
+bin_f.write_header({ 'eud': '-1', 'joiner': 'joiner.py', 'reward': 'latest'})
+
+for msg in interactions_file.messages():
```

Comment:
So a joined batch could have events that do not have observations in them, right?

Reply:
Yes, interactions without observations should be included in the output, it's a left outer join.